File: plugins/codex-usage-sidebar/native/Tests/SidebarCoreTests/AccountIdentityDecoderTests.swift
import Foundation
import XCTest
@testable import SidebarCore

final class AccountIdentityDecoderTests: XCTestCase {
    func testDecodesDisplayNameEmailAndAvatarFromAccountRead() throws {
        let data = Data(
            #"{"result":{"account":{"displayName":"Jace","email":"jace@example.com","avatarUrl":"https://example.com/avatar.png"}}}"#.utf8
        )

        let identity = try AccountIdentityDecoder.decodeResponse(data)

        XCTAssertEqual(identity.displayName, "Jace")
        XCTAssertEqual(identity.email, "jace@example.com")
        XCTAssertEqual(identity.avatarURL?.absoluteString, "https://example.com/avatar.png")
        XCTAssertEqual(identity.preferredName, "Jace")
    }

    func testFallsBackToEmailWhenAccountReadHasNoDisplayName() throws {
        let data = Data(
            #"{"result":{"account":{"email":"jace@example.com","planType":"plus"}}}"#.utf8
        )

        let identity = try AccountIdentityDecoder.decodeResponse(data)

        XCTAssertNil(identity.displayName)
        XCTAssertEqual(identity.preferredName, "jace@example.com")
    }

    func testRejectsAccountReadWithoutIdentityFields() {
        let data = Data(#"{"result":{"account":{"planType":"plus"}}}"#.utf8)

        XCTAssertThrowsError(try AccountIdentityDecoder.decodeResponse(data)) { error in
            XCTAssertEqual(error as? AccountIdentityDecodingError, .missingAccount)
        }
    }
}

File: plugins/codex-usage-sidebar/windows/tests/CodexUsageSidebar.Core.Tests/AccountIdentityDecoderTests.cs
using CodexUsageSidebar.Core;

namespace CodexUsageSidebar.Core.Tests;

[TestClass]
public sealed class AccountIdentityDecoderTests
{
    [TestMethod]
    public void AccountFieldsTakePrecedenceOverProfileAliases()
    {
        var account = AccountIdentityDecoder.DecodeResponse("""
            {"result":{"account":{"name":"Account name","image_url":"https://example.com/account.png",
              "profile":{"displayName":"Profile name","avatarUrl":"https://example.com/profile.png"}}}}
            """);
        Assert.AreEqual("Account name", account.DisplayName);
        Assert.AreEqual(new Uri("https://example.com/account.png"), account.AvatarUrl);
    }
    [TestMethod]
    public void DecodesDisplayNameEmailAndAvatarFromAccountResponse()
    {
        var identity = AccountIdentityDecoder.DecodeResponse(
            File.ReadAllText(ContractPath("account", "read-response.json")));

        Assert.AreEqual("Jace", identity.DisplayName);
        Assert.AreEqual("jace@example.com", identity.Email);
        Assert.AreEqual("https://example.com/avatar.png", identity.AvatarUrl?.ToString());
        Assert.AreEqual("Jace", identity.PreferredName);
    }

    [TestMethod]
    public void FallsBackToEmailWhenDisplayNameIsMissing()
    {
        var identity = AccountIdentityDecoder.DecodeResponse(
            "{\"result\":{\"account\":{\"email\":\"jace@example.com\"}}}");

        Assert.AreEqual("jace@example.com", identity.PreferredName);
    }

    private static string ContractPath(params string[] parts) =>
        Path.Combine(new[] { AppContext.BaseDirectory, "contracts" }.Concat(parts).ToArray());
}
